### ISSUE TITLE
View indexing is very slow, due to mis-optimized SQLite query

### ISSUE THREAD

**Issue by snej** [couchbase/couchbase-lite-ios#1737]
(CBSE-3690) View indexing has gotten drastically slower, at least in some use cases, in iOS 10 due to SQLite's mis-optimization of the main query used to find the new revisions to index — it takes a very long time (~500ms) and generates huge numbers of reads. 

Using `EXPLAIN QUERY PLAN` I found that the optimizer has generated a horrifically bad plan, essentially an inside-out join, that results in linear scans of two large tables. (Presumably this is due to some change in the optimizer between the older and newer versions of SQLite in iOS.)

The query, with some extraneous details removed, looks like:
```sql
	SELECT docs.doc_id, revs.sequence, docs.docid, rev.revid, rev.json, […] FROM revs, docs 
	WHERE revs.sequence>? AND revs.current!=0 AND revs.doc_id = docs.doc_id 
	ORDER BY revs.doc_id, revs.deleted, revs.revid DESC
```
Basically this query is meant to find the latest revs, including their “docid” strings joined from the “docs” table. It should just be using the “revs” table’s primary index starting from the “?” parameter, then looking up the corresponding doc for each revision. Instead it does this:
```
selectid = 0
  order = 0
   from = 1
 detail = SCAN TABLE docs USING COVERING INDEX docs_docid

selectid = 0
  order = 1
   from = 0
 detail = SEARCH TABLE revs USING INDEX revs_current (doc_id=?)

selectid = 0
  order = 0
   from = 0
 detail = USE TEMP B-TREE FOR ORDER BY
```
It’s scanning through all 50,000+ rows in “docs”, and then for each document looking up all the rows in “revs” that reference it (“revs” has 175,000 rows), and comparing their sequences to the parameter.

**Comment by pasin** [comment]
I'm flagging this as a cross platform issue. I will reopen the issue so that we can discuss during the sprint planning for porting the fix to the other platform.
